### ISSUE TITLE
CORS Error when session expires

### ISSUE THREAD

**Issue by nicolastakashi** [pomerium/pomerium#2176]
## What happened?

I have a page hosted at `mypage.corp.example.com` which makes an AJAX request to `api.corp.example.com`

When the AJAX call requests `api.corp.example.api.corp.example.`, pomerium redirects via 302 to: `https://auth.corp.example.com/.pomerium/sign_in?pomerium_expiry=12345&pomerium_issued=12345&pomerium_redirect_uri=https%3A%2F%2Fapi.corp.example.com&pomerium_signature=abcde`

The request made to `https://auth.corp.example.com/.pomerium/sign_in?pomerium_expiry=12345&pomerium_issued=12345&pomerium_redirect_uri=https%3A%2F%2Fapi.corp.example.com&pomerium_signature=abcde`

is returned with CORS Error!

I've configured CORS on Pomerium configuration but it does not work

## What did you expect to happen?

Refresh session without CORS issues

## What's your environment like?

- Pomerium version 0.11.1
- Server Operating System/Architecture/Cloud: Linux

## What's your config.yaml?

```config.yaml
- from: `api.corp.example`
        to: http://localhost:8080
        pass_identity_headers: true
        preserve_host_header: true
        cors_allow_preflight: true
        allowed_domains:
          - corm.example
```

## What did you see in the logs?
N/A

## Additional context
N/A

**Comment by nicolastakashi** [comment]
Any update guys?

**Comment by calebdoxsey** [comment]
Hi @nicolastakashi . I think I understand but even if we allowed CORS, it wouldn't fix your issue. You'd still see an HTML response with the login form, which the Javascript isn't going to understand. Furthermore we can't set the CORS header on the IDP login page, since that's not going through our proxy. (`sign_in` usually just redirects to the IDP)

Typically the way to refresh a session is to refresh the page so that the user can go through the login flow. This doesn't always play well with single-page apps, but I'm not sure of a better solution...

Does this sound like your issue? Or have I misunderstood?

**Comment by nicolastakashi** [comment]
@calebdoxsey yah, indeed refresh the page isn't the best solution ever and I really like to have a different solution :(

Since this kind of `bug` give us a bad user experience

**Comment by calebdoxsey** [comment]
@nicolastakashi Are you using the `fetch` API in javascript? If so I believe you can disable redirects and turn them into errors. You could then detect the error and refresh the page. Would that solve your problem?

**Comment by nicolastakashi** [comment]
@calebdoxsey I'm using pomerium in front of Grafana, as far as I know, Grafana is built on top of AngularJS

**Comment by calebdoxsey** [comment]
@nicolastakashi I see. Yes this is hard when you don't have control over the application. I'm not sure what a solution might look like. Some ideas:

1. Grafana plugins?
2. Maybe there's a way to install a `ServiceWorker` that can intercept the fetch request and cause a page reload?
3. Perhaps if the response is formatted in a certain way or has a certain response code Grafana has a built in way of forcing a reload? Presumably their builtin auth providers have mechanism for this...

This will require more investigation.

**Comment by calebdoxsey** [comment]
@nicolastakashi I just merged a change which makes the authorize service return a 401 instead of a redirect if `application/json` is the incoming accept header. Grafana has a special case for 401s which results in the login page being displayed, so I believe this should fix your issue.

**Comment by nicolastakashi** [comment]
Hey @calebdoxsey, sorry for the late reply.

First of all, thank you guys for the amazing work.

Is this fix available on version `v0.14.5`?

**Comment by calebdoxsey** [comment]
@nicolastakashi It's not available in `v0.14.5`, but I went ahead and backported it to the `0-14-0` branch and will go out with the next `v0.14` release.
